### PR TITLE
ARROW-4535: [C++] Fix MakeBuilder to preserve ListType's field name

### PR DESCRIPTION
**Pull request by mrkn**
MakeBuilder doesn't preserve the field name in the given ListType.
I think this is a bug.

### CODE REVIEW

**Review by wesm — APPROVED** [review]
+1, yes I agree with you

File: ruby/red-arrow/test/test-list-array.rb [hunk 1]
```diff
@@ -25,6 +25,7 @@ class ListArrayTest < Test::Unit::TestCase
         [false, true, false],
       ]
       array = Arrow::ListArray.new(data_type, values)
+      assert_equal(data_type, array.value_data_type)
```
Comment (kou) [inline ruby/red-arrow/test/test-list-array.rb]: I think that we don't need this assertion.
This is tests on C++ layer.

Comment (mrkn) [reply ruby/red-arrow/test/test-list-array.rb]: Deleted.

**Comment by kou** [comment]
The test failure on Travis CI is related to a `garrow_list_array_new()` bug.
It must be passed `GArrowFiled *` instead of using `arrow::list()` internally.

We need to fix this bug. The fix of this breaks backward compatibility but this is acceptable. Because normal users don't use `garrow_list_array_new()` directly.

**Comment by mrkn** [comment]
@kou I'll make another pull-request to modify `garrow_list_array_new()`.

**Comment by kou** [comment]
Thanks!

**Review by xhochy — APPROVED** [review]
+1, LGTM

**Pull request closed without merging**